src/group_theory/solvable.lean

```diff
@@ -0,0 +1,179 @@
+/-
+Copyright (c) 2021 Jordan Brown, Thomas Browning and Patrick Lutz. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jordan Brown, Thomas Browning and Patrick Lutz
+-/
+
+import group_theory.abelianization
+import data.bracket
+
+/-!
+# Solvable Groups
+
+In this file we introduce the notion of a solvable group. We define a solvable group as one whose
+derived series is eventually trivial. This requires defining the commutator of two subgroups and
+the derived series of a group.
+
+## Main definitions
+
+* `general_commutator H₁ H₂` : the commutator of the subgroups `H₁` and `H₂`
+* `derived_series G n` : the `n`th term in the derived series of `G`, defined by iterating
+  `general_commutator` starting with the top subgroup
+* `is_solvable G` : the group `G` is solvable
+-/
+
+open subgroup
+
+variables {G : Type*} [group G]
+
+section general_commutator
+
+/-- The commutator of two subgroups `H₁` and `H₂`. -/
+instance general_commutator : has_bracket (subgroup G) (subgroup G) :=
+⟨λ H₁ H₂, closure {x | ∃ (p ∈ H₁) (q ∈ H₂), p * q * p⁻¹ * q⁻¹ = x}⟩
+
+lemma general_commutator_def (H₁ H₂ : subgroup G) :
+  ⁅H₁, H₂⁆ = closure {x | ∃ (p ∈ H₁) (q ∈ H₂), p * q * p⁻¹ * q⁻¹ = x} := rfl
+
+instance general_commutator_normal (H₁ H₂ : subgroup G) [h₁ : H₁.normal]
+  [h₂ : H₂.normal] : normal ⁅H₁, H₂⁆ :=
+begin
+  let base : set G := {x | ∃ (p ∈ H₁) (q ∈ H₂), p * q * p⁻¹ * q⁻¹ = x},
+  suffices h_base : base = group.conjugates_of_set base,
+  { dsimp only [general_commutator_def, ←base],
+    rw h_base,
+    exact subgroup.normal_closure_normal },
+  apply set.subset.antisymm group.subset_conjugates_of_set,
+  intros a h,
+  rw group.mem_conjugates_of_set_iff at h,
+  rcases h with ⟨b, ⟨c, hc, e, he, rfl⟩, d, rfl⟩,
+  exact ⟨d * c * d⁻¹, h₁.conj_mem c hc d, d * e * d⁻¹, h₂.conj_mem e he d, by group⟩,
+end
+
+lemma general_commutator_mono {H₁ H₂ K₁ K₂ : subgroup G} (h₁ : H₁ ≤ K₁) (h₂ : H₂ ≤ K₂) :
+  ⁅H₁, H₂⁆ ≤ ⁅K₁, K₂⁆ :=
+begin
+  apply closure_mono,
+  rintros x ⟨p, hp, q, hq, rfl⟩,
+  exact ⟨p, h₁ hp, q, h₂ hq, rfl⟩,
+end
+
+lemma general_commutator_def' (H₁ H₂ : subgroup G) [H₁.normal] [H₂.normal] :
+  ⁅H₁, H₂⁆ = normal_closure {x | ∃ (p ∈ H₁) (q ∈ H₂), p * q * p⁻¹ * q⁻¹ = x} :=
+by rw [← normal_closure_eq_self ⁅H₁, H₂⁆, general_commutator_def,
+  normal_closure_closure_eq_normal_closure]
+
+lemma general_commutator_le (H₁ H₂ : subgroup G) (K : subgroup G) :
+  ⁅H₁, H₂⁆ ≤ K ↔ ∀ (p ∈ H₁) (q ∈ H₂), p * q * p⁻¹ * q⁻¹ ∈ K :=
+begin
+  rw [general_commutator, closure_le],
+  split,
+  { intros h p hp q hq,
+    exact h ⟨p, hp, q, hq, rfl⟩, },
+  { rintros h x ⟨p, hp, q, hq, rfl⟩,
+    exact h p hp q hq, }
+end
+
+lemma general_commutator_comm' (H₁ H₂ : subgroup G) : ⁅H₁, H₂⁆ ≤ ⁅H₂, H₁⁆ :=
```

Comment:
I think it might be nicer to "linline" this lemma within `general_commutator_comm` like what we have [here](https://github.com/leanprover-community/mathlib/blob/57c6d191e33800adc5229932aee69f334abd0bee/src/algebra/lie/basic.lean#L941).

After all, nobody will ever want to use `general_commutator_comm'` given that `general_commutator_comm` exists.

Reply:
Okay, done.